plutus-ledger-api/src/PlutusLedgerApi/Common/Eval.hs

```diff
@@ -249,10 +251,31 @@ evaluateScriptCounting ll pv verbose ectx p args = swap $ runWriter @LogOutput $
     appliedTerm <- mkTermToEvaluate ll pv p args
     let (res, UPLC.CountingSt final, logs) =
             evaluateTerm UPLC.counting pv verbose ectx appliedTerm
-    tell logs
-    liftEither $ first CekError $ void res
+    processLogsAndErrors ll logs res
     pure final
 
+processLogsAndErrors ::
+    forall m.
+    (MonadError EvaluationError m, MonadWriter LogOutput m) =>
+    PlutusLedgerLanguage ->
+    LogOutput ->
+    Either
+        (UPLC.CekEvaluationException NamedDeBruijn DefaultUni DefaultFun)
+        (UPLC.Term UPLC.NamedDeBruijn DefaultUni DefaultFun ()) ->
+    m ()
+processLogsAndErrors ll logs res = do
+    tell logs
+    case res of
+        Left e -> throwError (CekError e)
+        Right v ->
+            unless (ll == PlutusV1 || ll == PlutusV2 || isBuiltinUnit v) $
```

Comment:
this needs a Note! With a reference to the CIP

Reply:
I'd also be tempted to extract this out to a top-level predicate with a descriptive name and the explanation attached to it. It's a significant thing, worth making it look more significant!